REGRESSION(265569@main): SVG paths can get clipped at tile boundaries

shallawa opened the pull request:
#### 5eadead07f655bd2a9728cf87b8bbfe89e9e8cda
<pre>
REGRESSION(265569@main): SVG paths can get clipped at tile boundaries
<a href="https://bugs.webkit.org/show_bug.cgi?id=279105">https://bugs.webkit.org/show_bug.cgi?id=279105</a>
<a href="https://rdar.apple.com/135280080">rdar://135280080</a>

Reviewed by Simon Fraser.

When the distance between the current point and the end point of a cubic Bézier
curve is split into three equal parts with the two control points, the slope of
the curve can&apos;t be calculated from the derivative of the curve equation.

But in this case, the curve has to be quadratic which means the t-coordinate of
its vertex must be the mid-point between the current point and the end point.

* LayoutTests/svg/custom/quadratic-path-in-tiled-layer-expected.html: Added.
* LayoutTests/svg/custom/quadratic-path-in-tiled-layer.html: Added.
* Source/WebCore/platform/graphics/PathSegmentData.cpp:
(WebCore::calculateBezierExtremities):

Canonical link: <a href="https://commits.webkit.org/286550@main">https://commits.webkit.org/286550@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a334d155fe9fb73586b9cd06ceb305ab6159ab24

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76234 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55264 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29136 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80740 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27506 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78350 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64408 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3560 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59827 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17915 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79301 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49678 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65470 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40121 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47075 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22956 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25828 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68205 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23290 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82199 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3606 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2353 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68002 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3760 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65440 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67311 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16797 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11268 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9381 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3554 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6361 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3577 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7006 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5335 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/svg/custom/quadratic-path-in-tiled-layer.html:
```diff
@@ -0,0 +1,27 @@
+    <meta name="fuzzy" content="maxDifference=0-148; totalPixels=0-502" />
+    <style>
+        .container {
+            width: 600px;
```
Shouldn't this be much larger?

shallawa committed:
REGRESSION(265569@main): SVG paths can get clipped at tile boundaries
https://bugs.webkit.org/show_bug.cgi?id=279105
rdar://135280080

Reviewed by Simon Fraser.

When the distance between the current point and the end point of a cubic Bézier
curve is split into three equal parts with the two control points, the slope of
the curve can't be calculated from the derivative of the curve equation.

But in this case, the curve has to be quadratic which means the t-coordinate of
its vertex must be the mid-point between the current point and the end point.

* LayoutTests/svg/custom/quadratic-path-in-tiled-layer-expected.html: Added.
* LayoutTests/svg/custom/quadratic-path-in-tiled-layer.html: Added.
* Source/WebCore/platform/graphics/PathSegmentData.cpp:
(WebCore::calculateBezierExtremities):

Canonical link: https://commits.webkit.org/286550@main
